core/embed/extmod/modtrezorui/display.c

```diff
@@ -291,12 +255,48 @@ void display_image(int x, int y, int w, int h, const void *data,
     const int px = pos % w;
     const int py = pos / w;
     if (px >= x0 && px <= x1 && py >= y0 && py <= y1) {
-      PIXELDATA((decomp_out[0] << 8) | decomp_out[1]);
+      PIXELDATA((decomp_out[1] << 8) | decomp_out[0]);
     }
     decomp.dest = (uint8_t *)&decomp_out;
   }
 #endif
 }
+#else
+void display_image(int x, int y, int w, int h, const void *data,
+                   uint32_t datalen) {
+  x += DISPLAY_OFFSET.x;
+  y += DISPLAY_OFFSET.y;
+  int x0 = 0, y0 = 0, x1 = 0, y1 = 0;
+  clamp_coords(x, y, w, h, &x0, &y0, &x1, &y1);
+  display_set_window(x0, y0, x1, y1);
+  x0 -= x;
+  x1 -= x;
+  y0 -= y;
+  y1 -= y;
+
+  struct uzlib_uncomp decomp = {0};
+  uint8_t decomp_window[UZLIB_WINDOW_SIZE] = {0};
+
+  line_buffer_16bpp_t *b1 = buffers_get_line_buffer_16bpp(0, false);
+  line_buffer_16bpp_t *b2 = buffers_get_line_buffer_16bpp(1, false);
+
+  uzlib_prepare(&decomp, decomp_window, data, datalen, b1, w * 2);
+
+  dma2d_setup_16bpp();
+
+  for (int32_t pos = 0; pos < h; pos++) {
+    int32_t pixels = w;
+    line_buffer_16bpp_t *next_buf = (pos % 2 == 1) ? b1 : b2;
+    decomp.dest = next_buf->buffer;
+    decomp.dest_limit = next_buf->buffer + w * 2;
+    int st = uzlib_uncompress(&decomp);
+    if (st < 0) break;  // error
+    dma2d_wait_for_transfer();
+    dma2d_start(next_buf->buffer, (uint8_t *)DISPLAY_DATA_ADDRESS, pixels);
```

Comment:
it discards the volatile qualifier. 

since it is passed to HAL as uint32_t anyway it seems inconsequential, but maybe it would be more precise to keep the volatile in the dma2d_start and discard it when converting to uint32_t